// stryker-mocha-runner/src/StrykerMochaReporter.js
"use strict";
var test_runner_1 = require('stryker-api/test_runner');
var log4js = require('log4js');
var log = log4js.getLogger('StrykerMochaReporter');
var StrykerMochaReporter = (function () {
    function StrykerMochaReporter(runner) {
        this.runResult = {
            result: test_runner_1.TestResult.Error,
            testNames: [],
            succeeded: 0,
            failed: 0,
            timeSpent: 0,
            errorMessages: []
        };
        this.registerEvents(runner);
    }
    StrykerMochaReporter.prototype.registerEvents = function (runner) {
        var _this = this;
        runner.on('start', function () {
            _this.startDate = new Date();
            log.debug('Starting Mocha test run');
        });
        runner.on('pass', function (test) {
            _this.runResult.succeeded++;
            _this.runResult.testNames.push(test.fullTitle());
            log.debug("Test passed: " + test.fullTitle());
        });
        runner.on('fail', function (test, err) {
            _this.runResult.failed++;
            _this.runResult.testNames.push(test.fullTitle());
            _this.runResult.errorMessages.push(err.message);
            console.log(err.stack);
            log.debug("Test failed: " + test.fullTitle() + ". Error: " + err.message);
        });
        runner.on('end', function () {
            _this.runResult.result = test_runner_1.TestResult.Complete;
            _this.runResult.timeSpent = new Date().getTime() - _this.startDate.getTime();
            runner.runResult = _this.runResult;
            log.debug("Mocha test run completed: " + _this.runResult.succeeded + "/" + (_this.runResult.succeeded + _this.runResult.failed) + " passed");
        });
    };
    return StrykerMochaReporter;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = StrykerMochaReporter;
