// stryker/src/utils/parserUtils.js
"use strict";
var _ = require('lodash');
var esprima = require('esprima');
var escodegen = require('escodegen');
var esprimaOptions = {
    comment: true,
    loc: true,
    range: true,
    tokens: true,
};
function parse(code) {
    if (code === undefined) {
        throw new Error('Code parameter cannot be undefined');
    }
    var abstractSyntaxTree = esprima.parse(code, esprimaOptions);
    return abstractSyntaxTree;
}
exports.parse = parse;
;
function collectFrozenNodes(abstractSyntaxTree, nodes) {
    nodes = nodes || [];
    if (!_.isArray(abstractSyntaxTree) && _.isObject(abstractSyntaxTree) && abstractSyntaxTree.type && _.isUndefined(abstractSyntaxTree.nodeID)) {
        abstractSyntaxTree.nodeID = nodes.length;
        nodes.push(abstractSyntaxTree);
    }
    Object.freeze(abstractSyntaxTree);
    _.forOwn(abstractSyntaxTree, function (childNode, i) {
        if (childNode instanceof Object && !(childNode instanceof Array)) {
            collectFrozenNodes(childNode, nodes);
        }
        else if (childNode instanceof Array) {
            _.forEach(childNode, function (arrayChild) {
                if (arrayChild instanceof Object && !(arrayChild instanceof Array)) {
                    collectFrozenNodes(arrayChild, nodes);
                }
            });
        }
    });
    return nodes;
}
exports.collectFrozenNodes = collectFrozenNodes;
;
function generate(node) {
    return escodegen.generate(node);
}
exports.generate = generate;
;
