// stryker/src/utils/StrykerTempFolder.js
"use strict";
var fs = require('fs');
var path = require('path');
var mkdirp = require('mkdirp');
var fileUtils = require('./fileUtils');
var log4js = require('log4js');
var log = log4js.getLogger('fileUtils');
var baseTempFolder = path.join(process.cwd(), '.stryker-tmp');
var tempFolder = path.join(baseTempFolder, random().toString());
ensureFolderExists(baseTempFolder);
ensureFolderExists(tempFolder);
function createRandomFolder(prefix) {
    return ensureFolderExists(tempFolder + path.sep + prefix + random());
}
function random() {
    return Math.ceil(Math.random() * 10000000);
}
function ensureFolderExists(path) {
    if (!fileOrFolderExists(path)) {
        mkdirp.sync(path);
    }
    return path;
}
function fileOrFolderExists(path) {
    try {
        fs.lstatSync(path);
        return true;
    }
    catch (error) {
        return false;
    }
}
function writeFile(filename, data) {
    return new Promise(function (resolve, reject) {
        fs.writeFile(filename, data, { encoding: 'utf8' }, function (error) {
            if (error) {
                reject(error);
            }
            else {
                resolve();
            }
        });
    });
}
function copyFile(fromFilename, toFilename) {
    return new Promise(function (resolve, reject) {
        var readStream = fs.createReadStream(fromFilename, { encoding: 'utf8' });
        var writeStream = fs.createWriteStream(toFilename, { encoding: 'utf8' });
        readStream.on('error', reject);
        writeStream.on('error', reject);
        readStream.pipe(writeStream);
        readStream.on('end', function () { return resolve(); });
    });
}
function clean() {
    log.debug("Cleaning stryker temp folder " + baseTempFolder);
    return fileUtils.deleteDir(baseTempFolder);
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    createRandomFolder: createRandomFolder,
    writeFile: writeFile,
    copyFile: copyFile,
    ensureFolderExists: ensureFolderExists,
    clean: clean
};
