// stryker/src/mutators/RemoveConditionalsMutator.js
"use strict";
var esprima_1 = require('esprima');
var RemoveConditionalsMutator = (function () {
    function RemoveConditionalsMutator() {
        this.name = 'RemoveConditionals';
        this.types = [esprima_1.Syntax.DoWhileStatement, esprima_1.Syntax.IfStatement, esprima_1.Syntax.ForStatement, esprima_1.Syntax.WhileStatement, esprima_1.Syntax.ConditionalExpression];
    }
    RemoveConditionalsMutator.prototype.applyMutations = function (node, copy) {
        if (this.canMutate(node)) {
            var nodes = [];
            nodes.push(this.booleanLiteralNode(node.test.nodeID, false));
            if (node.type === esprima_1.Syntax.IfStatement || node.type === esprima_1.Syntax.ConditionalExpression) {
                nodes.push(this.booleanLiteralNode(node.test.nodeID, true));
            }
            return nodes;
        }
    };
    RemoveConditionalsMutator.prototype.booleanLiteralNode = function (nodeID, value) {
        return {
            nodeID: nodeID,
            type: esprima_1.Syntax.Literal,
            value: value,
            raw: value.toString()
        };
    };
    RemoveConditionalsMutator.prototype.canMutate = function (node) {
        return this.types.indexOf(node.type) >= 0;
    };
    ;
    return RemoveConditionalsMutator;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = RemoveConditionalsMutator;
