// stryker/src/MutantRunResultMatcher.js
"use strict";
var MutantRunResultMatcher = (function () {
    function MutantRunResultMatcher(mutants, runResultsByTestId) {
        this.mutants = mutants;
        this.runResultsByTestId = runResultsByTestId;
    }
    MutantRunResultMatcher.prototype.matchWithMutants = function () {
        var _this = this;
        this.mutants.forEach(function (mutant) {
            var smallestStatement;
            _this.runResultsByTestId.forEach(function (testResult, id) {
                var covered = false;
                if (testResult.coverage) {
                    var coveredFile = testResult.coverage[mutant.filename];
                    if (coveredFile) {
                        if (!smallestStatement) {
                            smallestStatement = _this.findSmallestCoveringStatement(mutant, coveredFile);
                        }
                        covered = coveredFile.s[smallestStatement] > 0;
                    }
                }
                else {
                    covered = true;
                }
                if (covered) {
                    mutant.addRunResultForTest(id, testResult);
                }
            });
        });
    };
    MutantRunResultMatcher.prototype.findSmallestCoveringStatement = function (mutant, coveredFile) {
        var _this = this;
        var smallestStatement;
        Object.keys(coveredFile.statementMap).forEach(function (statementId) {
            var location = coveredFile.statementMap[statementId];
            if (_this.statementCoversMutant(mutant, location) && _this.isNewSmallestStatement(coveredFile.statementMap[smallestStatement], location)) {
                smallestStatement = statementId;
            }
        });
        return smallestStatement;
    };
    MutantRunResultMatcher.prototype.isNewSmallestStatement = function (originalLocation, newLocation) {
        var statementIsSmallestStatement = false;
        if (!originalLocation) {
            statementIsSmallestStatement = true;
        }
        else {
            var lineDifference = (originalLocation.end.line - originalLocation.start.line) - (newLocation.end.line - newLocation.start.line);
            var coversLessLines = lineDifference > 0;
            var coversLessColumns = lineDifference === 0 && (newLocation.start.column - originalLocation.start.column) + (originalLocation.end.column - newLocation.end.column) > 0;
            if (coversLessLines || coversLessColumns) {
                statementIsSmallestStatement = true;
            }
        }
        return statementIsSmallestStatement;
    };
    MutantRunResultMatcher.prototype.statementCoversMutant = function (mutant, location) {
        var mutantIsAfterStart = mutant.location.start.line > location.start.line ||
            (mutant.location.start.line === location.start.line && mutant.location.start.column >= location.start.column);
        var mutantIsBeforeEnd = mutant.location.end.line < location.end.line ||
            (mutant.location.end.line === location.end.line && mutant.location.end.column <= location.end.column);
        return mutantIsAfterStart && mutantIsBeforeEnd;
    };
    return MutantRunResultMatcher;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = MutantRunResultMatcher;
